base/tuple.jl

```diff
@@ -31,6 +31,42 @@ function _setindex(v, i::Integer, first, tail...)
 end
 _setindex(v, i::Integer) = ()
 
+getindex(into::Tuple{}, index::Tuple{}) = ()
+getindex(into::Tuple, index::Tuple{}) = ()
+getindex(into::Tuple{}, index::Tuple) = ()
+getindex(into::Tuple, index::Tuple) = begin
+    next = getindex(tail(into), tail(index))
+    ifelse(first(index), (first(into), next...), next)
+end
+
+setindex(old::Tuple{}, new, switch::Tuple{}) = ()
+setindex(old::Tuple{}, new, switch::Tuple) = ()
+setindex(old::Tuple, new, switch::Tuple{}) = old
+setindex(old::Tuple, new, switch::Tuple) =  begin
+    first_switch = first(switch)
+    ifelse(first_switch, new, first(old)),
+    setindex(
+        tail(old),
+        new,
+        tail(switch))...
+end
+
+setindex(old::Tuple{}, new::Tuple{}, switch::Tuple{}) = ()
```

Comment:
I don't think this behavior (doing something different when the value to assign is a `Tuple`) fits well with how tuples work --- a tuple is equally likely to contain any kind of value, including another tuple. The similar method for assigning arrays (`array1[inds] = array2`) I think is also a wart, but less harmful in practice due to how arrays tend to be used.

Reply:
Ok. Is there a better name to use for this behavior? Seems useful.

Reply:
Now that we have nice broadcasting of tuples, I think this is the same thing as `ifelse.(mask, input, default)`, which already works and generates efficient code:
```julia
julia> foo(mask, input, default) = ifelse.(mask, input, default)
foo (generic function with 1 method)

julia> @code_warntype foo((false, true), (1, 2), 0)
Variables:
  mask::Tuple{Bool,Bool}
  input::Tuple{Int64,Int64}
  default::Int64

Body:
  begin
      # meta: location broadcast.jl broadcast 434
      # meta: location broadcast.jl broadcast_c 335
      # meta: location broadcast.jl tuplebroadcast 337
      SSAValue(5) = input::Tuple{Int64,Int64}
      SSAValue(6) = default::Int64
      # meta: pop location
      # meta: pop location
      # meta: pop location
      return (Core.tuple)((Base.select_value)((Base.getfield)(mask::Tuple{Bool,Bool}, 1, true)::Bool, (Base.getfield)(SSAValue(5), 1, true)::Int64, SSAValue(6))::Int64, (Base.select_value)((Base.getfield)(mask::Tuple{Bool,Bool}, 2, true)::Bool, (Base.getfield)(SSAValue(5), 2, true)::Int64, SSAValue(6))::Int64)::Tuple{Int64,Int64}
  end::Tuple{Int64,Int64}
```
I don't doubt that there are many places in Base that could be cleaned up by using this.

Reply:
Oh yeah, that is much nicer for methods of setindex where the replacement value is a scalar. Still need something for slotting in a tuple of replacement values?